File: php/ais_sdk/long_sentence.php
<?php
require "signer.php";
require "ais.php";

/**
 * token 方式
 * @return stri
 */
function long_sentence($token, $data, $url = "", $category = "common")
{
    // 获取任务信息
    $jobResult = _long_sentence($token, $data, $url, $category);
    $jobResultObj = json_decode($jobResult, true);
    $job_id = $jobResultObj['result']['job_id'];
    echo "Process job id is :" . $job_id;

    $words = "";
    while (true) {

        // 获取任务解析的结果
        $resultobj = get_result($token, $job_id);
        if ($resultobj['status'] != 200) {
            var_dump($resultobj);
        }

        if ($resultobj['result']['status_code'] == -1) {

            var_dump($resultobj);

        } // 任务处理完毕
        elseif ($resultobj['result']['status_code'] == 2) {

            $words = $resultobj['result']['words'];
            break;
        } // 任务未完成处理，轮询请求接口处理
        else {
            sleep(2);
            continue;
        }

    }
    return $words;

}


function _long_sentence($token, $data, $url = "", $category = "common")
{
    // 构建请求信息
    $_url = "https://" . ENDPOINT . LONG_SENTENCE;

    $data = array(
        "data" => $data,                        // data: 音频文件的base64
        "url" => $url,                          // url: 音频文件的obs的url地址
        "category" => $category,                // category：语音交流方式
    );

    $curl = curl_init();
    $headers = array(
        "Content-Type:application/json",
        "X-Auth-Token:" . $token);

    // 请求信息封装
    curl_setopt($curl, CURLOPT_URL, $_url);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($data));
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($curl, CURLOPT_NOBODY, FALSE);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

    // 执行请求信息
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {
            return $response;
        } else {
            echo "The http status code for get jobId request failure: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);
}

/**
 * 获取结果值
 * @param $token
 * @param $job_id
 */
function get_result($token, $job_id)
{
    $url = "https://" . ENDPOINT . LONG_SENTENCE . "?job_id=" . $job_id;

    $headers = array(
        "Content-Type:application/json",
        "X-Auth-Token:" . $token);

    $curl = curl_init(); // 启动一个CURL会话
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($curl, CURLOPT_HEADER, 0);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

    // 执行请求信息
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);

    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {

            $response = json_decode($response, true);
            $response['status'] = $status;
            return $response;
        } else {
            echo "The http status code for get result request failure: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);
}

/**
 * ak,sk 方式
 */
function long_sentence_aksk($_ak, $_sk, $data, $url = "", $category = "common")
{
    // 构建ak，sk对象
    $signer = new Signer();
    $signer->AppKey = $_ak;             // 构建ak
    $signer->AppSecret = $_sk;          // 构建sk

    $jobResult = _long_sentence_aksk($signer, $data, $url, $category);
    $jobResultObj = json_decode($jobResult, true);
    $job_id = $jobResultObj['result']['job_id'];
    echo "Process job id is :" . $job_id;

    $words = "";
    while (true) {

        // 获取任务的执行结果
        $resultobj = get_result_aksk($signer, $job_id);

        if ($resultobj['status'] != 200) {
            var_dump($resultobj);
        }

        if ($resultobj['result']['status_code'] == -1) {
            var_dump($resultobj);

            // 任务处理成功，返回结果信息
        } elseif ($resultobj['result']['status_code'] == 2) {
            $words = $resultobj['result']['words'];
            break;
        } // 任务处理未完成，轮询继续请求接口
        else {
            sleep(5);
            continue;
        }

    }
    return $words;
}

/**
 * 获取任务信息
 */
function _long_sentence_aksk($signer, $data, $url = "", $category = "common")
{
    // 构建请求信息
    $data = array(
        "data" => $data,                        // data: 音频文件的base64
        "url" => $url,                          // url: 音频文件的obs的url地址
        "category" => $category,                // category：语音交流方式
    );

    $req = new Request();
    $req->method = 'POST';
    $req->scheme = 'https';
    $req->host = ENDPOINT;
    $req->uri = LONG_SENTENCE;
    $req->body = json_encode($data);
    $req->headers = array(
        'Content-Type' => 'application/json'
    );
    $curl = $signer->Sign($req);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

    // 执行请求信息
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {
            return $response;
        } else {
            echo "The http status code for get jobId request failure: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);
}


/**
 * 获取结果值
 * @param $token
 * @param $job_id
 */
function get_result_aksk($signer, $job_id)
{

    $req = new Request();
    $req->method = 'GET';
    $req->scheme = 'https';
    $req->host = ENDPOINT;
    $req->uri = LONG_SENTENCE;
    $req->query = array(
        'job_id' => $job_id
    );
    $req->headers = array(
        'Content-Type' => 'application/json'
    );
    $curl = $signer->Sign($req);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

    // 执行请求信息
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);

    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {

            $response = json_decode($response, true);
            $response['status'] = $status;
            return $response;
        } else {
            echo "The http status code for get result request failure: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);
}

File: python/ais_sdk/long_sentence.py
# -*- coding:utf-8 -*-

import sys
import time
import json
import ais_sdk.ais as ais
import ais_sdk.utils as utils
import ais_sdk.signer as signer

#
# access asr, long_sentence,post data by token
#
def long_sentence(token, data, url=''):
    status, r = _long_sentence(token, data, url)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    #print "Process job id is :", job_id
    words = ''
    time.sleep(1.0)
    try:
        while True:
            status, r = _get_result(token, job_id)
            if status != 200:
                return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status_code', 1)
            if process_status == -1:
                return r

            elif process_status == 2:
                words = rec_result["result"].get('words', '')
                break

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''
    return words


#
# long_sentence, post the data
#
def _long_sentence(token, data, url):
    _url = 'https://%s/v1.0/voice/asr/long-sentence' % ais.AisEndpoint.ASR_ENDPOINT

    if sys.version_info.major >= 3:
        if data != '':
            data = data.decode("utf-8")
    _data = {
        "url": url,
        "data": data
    }

    status, resp = utils.request_token(_url, _data, token)
    if sys.version_info.major < 3:
        return status, resp
    else:
        return status, resp.decode('utf-8')


#
# access asr, long_sentence, get the result
#
def _get_result(token, job_id):
    _url_tmpl = 'https://%s/v1.0/voice/asr/long-sentence?job_id=%s'
    _url = _url_tmpl % (ais.AisEndpoint.ASR_ENDPOINT, job_id)
    return utils.request_job_result_token(_url, token)


#
# access asr, long_sentence,post data by ak,sk
#
def long_sentence_aksk(_ak, _sk, data, url=''):
    sig = signer.Signer()
    sig.AppKey = _ak
    sig.AppSecret = _sk

    status, r = _long_sentence_aksk(sig, data, url)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    #print "Process job id is :", job_id
    words = ''
    time.sleep(1.0)
    try:
        while True:
            status, r = _get_result_aksk(sig, job_id)
            if status != 200:
                return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status_code', 1)
            if process_status == -1:
                return r

            elif process_status == 2:
                words = rec_result["result"].get('words', '')
                break

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''
    return words


#
# long_sentence, post the data
#
def _long_sentence_aksk(sig, data, url):
    _url = 'https://%s/v1.0/voice/asr/long-sentence' % ais.AisEndpoint.ASR_ENDPOINT

    if sys.version_info.major >= 3:
        if data != '':
            data = data.decode("utf-8")

    _data = {
        "url": url,
        "data": data
    }

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = ais.AisEndpoint.ASR_ENDPOINT
    kreq.uri = "/v1.0/voice/asr/long-sentence"
    kreq.method = "POST"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.body = json.dumps(_data)

    status, resp = utils.request_aksk(sig, kreq, _url)
    if sys.version_info.major < 3:
        return status, resp
    else:
        return status, resp.decode('utf-8')


#
# access asr, long_sentence, get the result
#
def _get_result_aksk(sig, job_id):
    _url_tmpl = 'https://%s/v1.0/voice/asr/long-sentence?job_id=%s'
    _url = _url_tmpl % (ais.AisEndpoint.ASR_ENDPOINT, job_id)

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = ais.AisEndpoint.ASR_ENDPOINT
    kreq.uri = "/v1.0/voice/asr/long-sentence"
    kreq.method = "GET"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.query = {'job_id': job_id}

    return utils.request_job_result_aksk(sig, kreq, _url)
